layout-creation.md

```diff
@@ -0,0 +1,99 @@
+# Layout Creation Example
+
+The following Python snippet shows how to use in-toto to create a software
+supply chain layout like the one that is used for the in-toto demo.
+*Take a look at the [demo repo](https://github.com/in-toto/demo) for more
+information about that supply chain*.
+
+
+```python
+from securesystemslib.interface import (generate_and_write_rsa_keypair,
+    import_rsa_privatekey_from_file)
+from in_toto.models.layout import Layout, Step, Inspection
+from in_toto.models.metadata import Metablock
+
+# Create RSA private and public keys
+
+# Alice is the project owner and her private key is used to sign the layout
+alice_path = generate_and_write_rsa_keypair("alice", password="123")
+alice_key = import_rsa_privatekey_from_file(alice_path, password="123")
+
+# Bob and Carl are both functionaries, i.e. they are authorized to carry out
+# different steps of the supply chain. Their public keys are added to the
+# supply chain layout, in order to verify the signatures of the link metadata
+# that Bob and Carl will generate when carrying out their respective tasks.
+bob_path = generate_and_write_rsa_keypair("bob", password="123")
+carl_path = generate_and_write_rsa_keypair("carl", password="123")
+
+
+# Create an empty layout
+layout = Layout()
+
+# Add functionary public keys to the layout
+bob_pubkey = layout.add_functionary_key_from_path(bob_path + ".pub")
+carl_pubkey = layout.add_functionary_key_from_path(carl_path + ".pub")
+
+# Set expiration date so that the layout will expire in 4 months
+layout.set_relative_expiration(months=4)
+
+
+# Create layout steps
+
+# Each step describes a task that is required to be carried out in this supply
+# chain. A step must have a unique name to associate the related link metadata
+# (i.e. the evidence that the step was carried out). Additionally, each step
+# should list rules, about the present related files before and after the step
+# was carried out. A steps pubkeys field lists the keyids of functionaries
+# authorized to perform the step.
+step_clone = Step(name="clone")
+step_clone.set_expected_command_from_string(
+    "git clone https://github.com/in-toto/demo-project.git")
+step_clone.add_product_rule_from_string("CREATE demo-project/foo.py")
+step_clone.add_product_rule_from_string("DISALLOW *")
+step_clone.pubkeys = [bob_pubkey["keyid"]]
+
+step_update = Step(name="update-version")
+step_update.add_material_rule_from_string(
+    "MATCH demo-project/* WITH PRODUCTS FROM clone")
+step_update.add_material_rule_from_string("DISALLOW *")
+step_update.add_product_rule_from_string("ALLOW demo-project/foo.py")
+step_update.add_product_rule_from_string("DISALLOW *")
```

Comment:
> step_update.add_product_rule_from_string("ALLOW demo-project/foo.py")
> step_update.add_product_rule_from_string("DISALLOW *")

These two rules together seem to contradict each other.  I personally find it easier to grok `DISALLOW * EXCEPT "demo-project/foo.py"` or `ALLOW * EXCEPT "foo"`.  However, I'm not familiar with in-toto's rules and why things are designed this way.

Is there a reason why `DISALLOW *` isn't the default case/behavior?

Reply:
(1) We can consider that. Care for opening an issue, maybe on the [specification repo](https://github.com/in-toto/docs)?
(2) We used to have implicit `DISALLOW *` but received a request to change it, see #131 and https://github.com/in-toto/docs/issues/4